src/transforms/key_value_parser.rs

```diff
@@ -0,0 +1,405 @@
+            "foo bar, beep bop, score 10",
+            ",".to_string(),
+            " ".to_string(),
+            false,
+            &[],
+            None,
+            None,
+            None,
+        );
+        assert_eq!(log[&"foo".into()], Value::Bytes("bar".into()));
+        assert_eq!(log[&"beep".into()], Value::Bytes("bop".into()));
+    }
+
+    #[test]
+    fn it_handles_whitespace_in_fields() {
+        let log = parse_log(
+            "foo:bar, beep : bop, score :10",
+            ",".to_string(),
+            ":".to_string(),
+            false,
+            &[("score", "integer")],
+            None,
+            None,
+            None,
+        );
+        assert_eq!(log[&"foo".into()], Value::Bytes("bar".into()));
+        assert_eq!(log[&"beep".into()], Value::Bytes("bop".into()));
+        assert_eq!(log[&"score".into()], Value::Integer(10));
+    }
+
+    #[test]
+    fn it_handles_multi_char_splitters() {
+        let log = parse_log(
+            "foo=>bar || beep => bop || score=>10",
+            "||".to_string(),
+            "=>".to_string(),
+            false,
+            &[("score", "integer")],
+            None,
+            None,
+            None,
+        );
+
+        assert_eq!(log[&"foo".into()], Value::Bytes("bar".into()));
+        assert_eq!(log[&"beep".into()], Value::Bytes("bop".into()));
+        assert_eq!(log[&"score".into()], Value::Integer(10));
+    }
+
+    #[test]
+    fn it_handles_splitters_in_value() {
+        let log = parse_log(
+            "foo==bar, beep=bop=bap , score=10",
+            ",".to_string(),
+            "=".to_string(),
+            false,
+            &[("score", "integer")],
+            None,
+            None,
+            None,
+        );
+        assert_eq!(log[&"foo".into()], Value::Bytes("bar".into()));
+        assert_eq!(log[&"beep".into()], Value::Bytes("bop=bap".into()));
+        assert_eq!(log[&"score".into()], Value::Integer(10));
+    }
+
+    #[test]
+    fn it_fails_graceful_on_empty_values() {
+        let log = parse_log(
+            "foo::0, ::beep, score:: ",
+            ",".to_string(),
+            "::".to_string(),
+            false,
+            &[],
+            None,
+            None,
+            None,
+        );
+        assert!(log.contains(&"foo".into()));
+        assert!(!log.contains(&"beep".into()));
+        assert!(!log.contains(&"score".into()));
+    }
+
+    #[test]
+    fn it_accepts_brackets() {
+        let log = parse_log(
+            "{\"foo\"}:0, \"\"bop\":[beep], [score]:78",
+            ",".to_string(),
+            ":".to_string(),
+            false,
+            &[],
+            None,
+            Some("\"{}".to_string()),
+            None,
+        );
+
+        // println!("{}", log.keys().fold("".to_string(), |res, key| [res, key].join(", ")));
+        // Review: This fails and I don't know why. The key is missing from the results.
+        // assert!(log.contains(&"[score]".into()));
```

Comment:
@bruceg Any ideas why `[` chars would cause problems? As far as I can tell, the parser I've written should be okay with this char. This was a very weird problem. I think I spent an hour or two and couldn't figure out why it would strip keys with`[]`.

Reply:
Our internal event field name parsing uses `[X]` to mean an index on an array element, and so is trying to handle `[score]` as that and failing.

Reply:
I've removed this test.